stages/dracut.conf: add new stage for creating dracut config files

thozza opened the pull request:
Add new `org.osbuild.dracut.conf` stage allowing to create dracut configuration files in `/usr/lib/dracut/dracut.conf.d` and thus make the configuration persistent. The stage supports only a subset of all configuration options allowed in dracut configuration. The intention is to provide almost functional parity with the options supported by `org.osbuild.dracut` stage.

The schema mandates that at least one configuration file must be defined in the stage options. In addition, each configuration file must contain at least one configuration option.

Add test for the new stage.

gicmo commented on stages/org.osbuild.dracut.conf:
```diff
@@ -0,0 +1,184 @@
+"additionalProperties": false,
+"minProperties": 1,
+"patternProperties": {
+  "^[\\w.-]{1,250}\\.conf$": {
```
I wonder, as with the `chmod` stage, if we should not move this into a one sub level, to allow us the flexibility to have other global options that we might want in the future (like the config dir path). Alternatively we could have:

```
{
  "filename": "",
  "config": {
    "compress": "..."
  }
}
```

And if you want more than one file, you add another `dracut.conf` stage. I think I personally prefer that since it is a tiny bit simpler, but OTOH we have done it differently now in some other stages.

thozza replied:
Right. I think this is a matter of preference, but I'm all for consistency if we decide which approach should be preferred. The reason why I'm choosing this approach is that it is more convenient from my point of view while iterating over all the configuration files and processing them. However that is really very minor "advantage" which is also still subjective.

As for moving the schema by one sub level, I commented in another comment below...

gicmo replied:
Yeah, I agree, especially on the consistency domain. I _feel_ (hehe) having one stage write one feel is somehow nicer, makes the schema and the stage itself minimally easier (one less indirection) and even on the composer side one less nested struct. Hence my personal preference (`configuration_files` also seems verbose and generic, i.e. vs. `users` in `org.osbuild.user` top level).

But for consistency, maybe we should have the `configuration_files` key for now, and then change it in all the new stages in one PR, _if_ we decide for that.

gicmo commented on stages/org.osbuild.dracut.conf:
```diff
@@ -0,0 +1,184 @@
+
+    # options with format like:
+    # persistent_policy="<policy>"
+    SIMPLE_STR_OPTIONS = [
+        "compress"
+    ]
+    # options with format like:
+    # add_dracutmodules+=" <dracut modules> "
+    ADD_LIST_OPTIONS = [
+        "add_dracutmodules",
+        "dracutmodules",
+        "omit_dracutmodules",
+        "drivers",
+        "add_drivers",
+        "force_drivers",
+        "filesystems",
+        "install_items"
+    ]
+    # options with format like:
+    # reproducible="{yes|no}"
+    BOOL_OPTOINS = [
+        "early_microcode",
+        "reproducible"
+    ]
```
I think I would have inverted this and have `option: type`, or even `option: writer_function`, but this is fine too.

thozza replied:
hmm... I'm not sure I understand what do you mean here. Do you mean that you would extend the schema and specify a type/writer_function for each configuration option?

gicmo replied:
Sorry, was not clear enough
```python
KNOWN_OPTIONS: {
   "compress": "string",
   "add_dracutmodules": "list",
   "early_microcode": "boolean"
}


for option, value in options.items():
   otype = KNOWN_OPTIONS[option]
   if otype == "string":
     f.write(f'{option}="{value}"\n')
   ...

```

or

```python

def string_writer(f, option, value):
    f.write(f'{option}="{value}"\n')

def list_writer(f, option, value):
  ...

KNOWN_OPTIONS: {
   "compress": string_writer,
   "add_dracutmodules": list_writer,
   "early_microcode": bool_writer
}


for option, value in options.items():
   writer = KNOWN_OPTIONS[option]
   writer(f, option, value)
```

The `KNONW_OPTIONS` thing is a bit funny since it basically is already there, in the `SCHEMA`, so we could in theory parse it and use the info form there too.

thozza replied:
Right. This makes sense and in most cases it will have lower time complexity. I'll rework it...

gicmo commented on stages/org.osbuild.dracut.conf:
```diff
@@ -0,0 +1,184 @@
+    with open(f"{config_files_dir}/{filename}", "w") as f:
+        for option, value in options.items():
+            if option in SIMPLE_STR_OPTIONS:
+                f.write(f"{option}=\"{value}\"\n")
```
I think this is the time where we can use `'` to avoid having to escape:
```suggestion
                f.write(f'{option}="{value}"\n')
```

gicmo reviewed (commented):
Looks good to me! Only small comments, could also keep it as is. Only thing I am slightly concerned about is that the current schema doesn't allow us to add any top level option in the future (which we might never need, but who knows), see inline comment.

thozza commented:
> Looks good to me! Only small comments, could also keep it as is. Only thing I am slightly concerned about is that the current schema doesn't allow us to add any top level option in the future (which we might never need, but who knows), see inline comment.

Yes, I took this approach intentionally, because the stage is already specific only to dracut configuration files. One thing that I also considered is a case when we would like to allow choosing the location of the configuration file (different than the default `/usr/lib/dracut/dracut.conf.d`. However this can be done by adding a property to the configuration file object, defaulting to the current path. So it could be specified per configuration file if needed.

Nevertheless I really don't feel strongly about keeping it as is and I'm happy to move the current schema under e.g. `configuration_files` property, similarly as it is done in `org.osbuild.modprobe` stage.

gicmo commented:
> > Looks good to me! Only small comments, could also keep it as is. Only thing I am slightly concerned about is that the current schema doesn't allow us to add any top level option in the future (which we might never need, but who knows), see inline comment.
> 
> Yes, I took this approach intentionally, because the stage is already specific only to dracut configuration files. One thing that I also considered is a case when we would like to allow choosing the location of the configuration file (different than the default `/usr/lib/dracut/dracut.conf.d`. However this can be done by adding a property to the configuration file object, defaulting to the current path. So it could be specified per configuration file if needed.

Yeah totally, but then we are mixing options _for_ the config file and options going _into_ the config file, which might be confusing.

> Nevertheless I really don't feel strongly about keeping it as is and I'm happy to move the current schema under e.g. `configuration_files` property, similarly as it is done in `org.osbuild.modprobe` stage.

We have done that for `groups` and `users` as well where even though the stage is already called `org.osbuild.user` we have `users` to top-level group and I am quite happy because I am now adding ostree support to the stage and thus need an `ostree` top level group.

thozza commented:
> We have done that for `groups` and `users` as well where even though the stage is already called `org.osbuild.user` we have `users` to top-level group and I am quite happy because I am now adding ostree support to the stage and thus need an `ostree` top level group.

True. Also the lack of a top-level group in the `org.osbuild.dracut` is the main reason why I'm adding this as a separate stage 🤣  (although is also makes sense from from the point of view of separating a bit different functionality - running dracut vs. configuring it)

thozza commented:
I updated the PR. Changes:

- Moved the original schema by a sub-level under `configuration_files` option.
- Modified some strings to use single quote to not have to escape double quotes.
- Reworked how options are processed and written to the configuration file using "writer" functions.

gicmo reviewed (approved):
👍 

Still not a fan of the `configration_files` thing, but I either get over it or we change it later.
